cabalscan/src/CabalScan/RuleGenerator.hs

```diff
@@ -214,6 +223,19 @@ generateRule cabalFilePath pkgId dataFiles bi someModules ctype attrName privAtt
     toToolName (Cabal.ExeDependency pkg exe _) =
       ToolName (pkgNameToText pkg) (Text.pack $ Cabal.unUnqualComponentName exe)
 
+mainFileAttr :: Path b File -> FilePath -> Cabal.BuildInfo -> IO Attributes
```

Comment:
Thanks @kczulko! Can I have a specification of this function?

Reply:
I saw that build is still failing. @facundominguez do you have ability to update the secret there https://github.com/tweag/gazelle_cabal/blob/main/.github/workflows/workflow.yaml#L27 ?

Reply:
Thanks for the explanation. I'd be more explicit in the type signature
```Haskell
mainFileAttr :: Path b File -> FilePath -> [FilePath] -> IO FilePath
```
The source dirs are passed to the function, instead of the large `BuildInfo`. And the function returns the path to the main file (relative to the cabal file).

Specifying the function in a comment would be necessary. A specification would say what the input parameters are, and how the output relates to them.

Also, this function isn't failing if the file is not found. It probably should produce an error in that case. You could use `findModulePaths` (plural) which does produce an error, I think, or take inspiration from it.

Reply:
> I saw that build is still failing.

This is because your fork doesn't have the secret set. I've pushed your commits to a branch in this repo. You can access the secret at https://app.buildbuddy.io/docs/setup/, but you need to login to buildbuddy first or otherwise that link will only show you generic instructions.

Reply:
> Also, this function isn't failing if the file is not found. It probably should produce an error in that case. You could use findModulePaths (plural) which does produce an error, I think, or take inspiration from it.

I'm not fully convinced about the failure here. There are at least two reasons why I think so:
- if something goes wrong during obtaining Main module filepath, gazelle_cabal would be still able to produce `BUILD.bazel` files instead of failing at the `cabalscan`
- there are some cases e.g. for benchmarks and tests where main file is represented through ADTs with constructors like [Cabal.TestSuiteExeV10](https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/Distribution-Types-TestSuiteInterface.html#v:TestSuiteExeV10) or [Cabal.BenchmarkExeV10](https://hackage.haskell.org/package/Cabal-3.6.3.0/docs/Distribution-Types-BenchmarkInterface.html#v:BenchmarkExeV10). Those data types contain also constructors which indicates failures like this:
> BenchmarkUnsupported BenchmarkType | A benchmark that does not conform to one of the above  interfaces for the given reason (e.g. unknown benchmark type).

Frankly speaking, I'm not aware for which cases cabal parser produces such values but for this scenario I feel more comfortable (feel more safety) when I put the data into the list and then **_eventually_** create `main_file`.

WDYT?

Reply:
> if something goes wrong during obtaining Main module filepath, gazelle_cabal would be still able to produce BUILD.bazel files instead of failing at the cabalscan

I think the earlier we fail, the better, or the user will have to rerun more than one tool after discovering the error.

> I'm not aware for which cases cabal parser produces such values but for this scenario

I don't know either. But I wouldn't try to anticipate it either. When we encounter those, we could consider what to do.